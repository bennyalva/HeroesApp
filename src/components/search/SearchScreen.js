import React, { useMemo } from 'react';
import queryString from 'query-string' ;
import { HeroCard } from '../heroes/HeroCard';
import { useForm } from '../../hooks/useForm';
import { useLocation } from 'react-router-dom';
import { getHeroesByName } from '../selectors/getHeroesByName';

export const SearchScreen = ({history}) => {
    const location = useLocation();
    const { q = ''} = queryString.parse( location.search);
    const [{heroeDesc}, handlerInputChange, reset ] = useForm({heroeDesc: q});
    const heroesFilters = useMemo(() => getHeroesByName( q ), [q])
    const handlerSubmit = (e) => {
        e.preventDefault();
        if (heroeDesc.trim().length <= 1 ) {
            return;
        }
        history.push(`?q=${ heroeDesc }`)
        console.log('heroe: ', heroeDesc)
    }
    return (
        <div>
            <h1>Search Screen</h1>
            <hr />
            <div className="row">
                <div className="col-5">
                    <h4>Search Form</h4>
                    <form onSubmit={handlerSubmit}>
                        <input
                            type="text"
                            name="heroeDesc"
                            placeholder="Find a heroe"
                            className="form-control" 
                            onChange={handlerInputChange}
                            value={heroeDesc}
                        />
                        <button 
                                type="submit"
                                className="btn m-1 btn-block btn-outline-primary">
                                    Search

                        </button>
                    </form>
                </div>
                <div className="col-7">
                    <h4> Results</h4>
                    {
                            (q === '')
                            &&
                            <div className="alert alert-info">
                                Search a heroe
                            </div>
                    }
                     {
                            (q !== '' && heroesFilters.length === 0)
                            &&
                            <div className="alert alert-danger">
                                There is no a heroe with {q}
                            </div>
                    }
                    {
                        heroesFilters.map(heroe => (
                            <HeroCard key={heroe.id}
                                      {...heroe} />
                        ))
                    }
                    <hr />
                </div>
            </div>
            
        </div>
    )
}
