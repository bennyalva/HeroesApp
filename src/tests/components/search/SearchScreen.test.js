import React from 'react';
import { mount } from 'enzyme';
import { SearchScreen } from '../../../components/search/SearchScreen';
import { MemoryRouter, Route } from 'react-router-dom';
describe('Test on Search screen', () => {
  
   test('should show correctly with default values', () => {
    const wrapper = mount(
        <MemoryRouter initialEntries={['/search']}>
            <Route path="/search" component={ SearchScreen } />
        </MemoryRouter>
    );
      expect(wrapper).toMatchSnapshot(); 
      expect(wrapper.find('.alert-info').text().trim()).toBe('Search a heroe');
   });
   test('should show batman and input with value of querystring', () => {
    const wrapper = mount(
        <MemoryRouter initialEntries={['/search?q=batman']}>
            <Route path="/search" component={ SearchScreen } />
        </MemoryRouter>
    );
        expect(wrapper.find('input').prop('value')).toBe('batman');
        expect(wrapper).toMatchSnapshot();
   });

   test('should show error with not foud heroe', () => {
    const wrapper = mount(
        <MemoryRouter initialEntries={['/search?q=batman123']}>
            <Route path="/search" component={ SearchScreen } />
        </MemoryRouter>
    );
    expect(wrapper.find('.alert-danger').text().trim()).toBe('There is no a heroe with batman123');
   
  });

  test('should call push on history', () => {
      const historyMock = {
          push: jest.fn()
      }
      const wrapper = mount(
        <MemoryRouter initialEntries={['/search?q=batman123']}>
            <Route path="/search" component={ () => <SearchScreen history={ historyMock} /> } />
        </MemoryRouter>
       );
       wrapper.find('input').simulate('change', {
           target: {
               name: 'heroeDesc',
               value: 'batman'
           }
       });
       wrapper.find('form').prop('onSubmit')({
           preventDefault(){}
       });
       expect(historyMock.push).toHaveBeenCalledWith('?q=batman')
  });
   
   
    
});
