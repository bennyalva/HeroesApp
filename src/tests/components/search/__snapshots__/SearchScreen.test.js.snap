// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test on Search screen should show batman and input with value of querystring 1`] = `
<div>
  <h1>
    Search Screen
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Search Form
      </h4>
      <form
        onSubmit={[Function]}
      >
        <input
          className="form-control"
          name="heroeDesc"
          onChange={[Function]}
          placeholder="Find a heroe"
          type="text"
          value="batman"
        />
        <button
          className="btn m-1 btn-block btn-outline-primary"
          type="submit"
        >
          Search
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
         Results
      </h4>
      <div
        className="card ms-3"
        style={
          Object {
            "maxWidth": 540,
          }
        }
      >
        <div
          className="row no-gutters"
        >
          <div
            className="col-md-4"
          >
            <img
              alt="Batman"
              className="card-img"
              src="./assets/heroes/dc-batman.jpg"
            />
          </div>
          <div
            className="col-md-8"
          >
            <div
              className="card-body"
            >
              <h5
                className="card-title"
              >
                Batman
              </h5>
              <p
                className="card-text"
              >
                Bruce Wayne
              </p>
              <p
                className="card-text"
              >
                <small
                  className="text-muted"
                >
                  Detective Comics #27
                </small>
              </p>
              <a
                href="/hero/dc-batman"
                onClick={[Function]}
              >
                MÃ¡s...
              </a>
            </div>
          </div>
        </div>
      </div>
      <hr />
    </div>
  </div>
</div>
`;

exports[`Test on Search screen should show correctly with default values 1`] = `
<div>
  <h1>
    Search Screen
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Search Form
      </h4>
      <form
        onSubmit={[Function]}
      >
        <input
          className="form-control"
          name="heroeDesc"
          onChange={[Function]}
          placeholder="Find a heroe"
          type="text"
          value=""
        />
        <button
          className="btn m-1 btn-block btn-outline-primary"
          type="submit"
        >
          Search
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
         Results
      </h4>
      <div
        className="alert alert-info"
      >
        Search a heroe
      </div>
      <hr />
    </div>
  </div>
</div>
`;
